src/workerd/io/actor-sqlite.c++

```diff
@@ -169,11 +201,83 @@ void ActorSqlite::onWrite() {
+  PrecommitAlarmState state;
+  state.localAlarmState = metadata.getAlarm();
+  if (pendingCommit == kj::none &&
+      willFireEarlier(state.localAlarmState, lastConfirmedScheduledAlarm)) {
+    // Basically, this is the first scheduling request that commitImpl() would make prior to
+    // commitCallback().  We start the request separately, ahead of calling sqlite functions that
+    // commit to local disk, for correctness in workerd, where alarm scheduling and db commits are
+    // both synchronous.
+    state.schedulingPromise = requestScheduledAlarm(state.localAlarmState);
+  }
+  return kj::mv(state);
+}
+
+kj::Promise<void> ActorSqlite::commitImpl(ActorSqlite::PrecommitAlarmState precommitAlarmState) {
+  // We assume that exceptions thrown during commit will propagate to the caller, such that they
+  // will ensure cancelDeferredAlarmDeletion() is called, if necessary.
+
+  KJ_IF_SOME(pending, pendingCommit) {
+    // If an earlier commitImpl() invocation is already in the process of updating precommit
+    // alarms but has not yet made the commitCallback() call, it should be OK to wait on it to
+    // perform the precommit alarm update and db commit for this invocation, too.
+    co_await pending.addBranch();
+    co_return;
+  }
+  auto [promise, fulfiller] = kj::newPromiseAndFulfiller<void>();
+  pendingCommit = promise.fork();
+
+  // Wait for the first precommit alarm scheduling request to complete, if any.
+  auto localAlarmState = precommitAlarmState.localAlarmState;
+  KJ_IF_SOME(p, precommitAlarmState.schedulingPromise) {
+    co_await p;
+    localAlarmState = metadata.getAlarm();
+  }
+
+  // While the local db state requires an earlier alarm than is known to be scheduled, issue an
+  // alarm update request for the earlier time.  This helps ensure that the successfully scheduled
+  // alarm time is always earlier or equal to the alarm state in the successfully persisted db.
+  while (willFireEarlier(localAlarmState, lastConfirmedScheduledAlarm)) {
+    co_await requestScheduledAlarm(localAlarmState);
+    localAlarmState = metadata.getAlarm();
+  }
+
+  // Issue the commitCallback() request to persist the db state, then synchronously clear the
+  // pending commit so that the next commitImpl() invocation starts its own set of precommit alarm
+  // updates and db commit.
+  auto commitCallbackPromise = commitCallback();
+  pendingCommit = kj::none;
+
+  // Wait for the db to persist.
+  co_await commitCallbackPromise;
+  lastConfirmedAlarmDbState = localAlarmState;
+
+  // Notify any merged commitImpl() requests that the db update completed.
+  fulfiller->fulfill();
+
+  // If the db state is now later than the known-scheduled alarm, issue a request to update it to
+  // match the db state.  We don't need to hold open the output gate, so we add the scheduling
+  // request to commitTasks.
+  if (willFireEarlier(lastConfirmedScheduledAlarm, localAlarmState)) {
```

Comment:
So IIUC:

1. We'll update the scheduled alarm time synchronouslybefore committing if we're moving it earlier
2. We'll update the scheduled alarm time asynchronously after committing if we're moving it later

But what if we do one and then the other, before we get back the result from the first request? Imagine something like:

1. Alarm scheduled/confirmed at time T
2. Set alarm to time T+100, which commits and kicks off a background request to move the alarm time back
3. Set alarm time to T+5 *before* the background request resolves, such that `lastConfirmedScheduledAlarm` is still set to T.
4. Since `lastConfirmedScheduledAlarm` was T, and T is earlier than T+5, we decide to update the alarm *after* committing.
5. Wait on `commitCallbackPromise`
6. The background request to move the alarm time back succeeds, updating `lastConfirmedScheduledAlarm` to T+100.
7. `commitCallbackPromise` resolves, but now T+5 *is* earlier than `lastConfirmedScheduledAlarm`, so this check evaluates to false and we end up never scheduling the alarm for T+5.

Am I missing something or could this leave us with an alarm that won't fire until T+100 even though it should fire at T+5?

Reply:
Wow... good catch.  I was worried there might be cases like this, but hadn't been able to find one.

I'm not sure of the best way to resolve the issue... I think we may need to have `startPrecommitAlarmScheduling()/commitImpl()` block on the completion of all outstanding scheduling requests prior to the initial `willFireEarlier()` check (while also, of course, ensuring that the first scheduling request is still issued synchronously in `startPrecommitAlarmScheduling()` for correctness in workerd).  I think that will limit the throughput of commits in flight when some contain alarm changes, but should otherwise work correctly.

That will also mostly limit scheduling requests to having one in-flight at a time, so it would probably make sense to explicitly enforce that as well, to eliminate the possibility of other races between the final alarm scheduling in `commitImpl()`.  Per Kenton's other comment, though, I think we can allow the output gate to remain open for that request.

@kentonv Does that make sense?

Reply:
Although if we're too tight on time before when this needs to ship, it wouldn't be disastrous to ship what you have and occasionally have delayed alarms in that kind of edge case for a week or two while we get a better fix in.

Reply:
I think that when making the post-commit request to move the alarm time _later_, you can _immediately_ update `lastConfirmedScheduledAlarm` to the new value when you kick off the request, without actually waiting for confirmation. This works because it's always OK if the alarm time accidentally gets left being too early -- you'll just reschedule it when it fires prematurely.

Reply:
Yeah, I think that works.

Although I still worry about whether we need to serialize the `hooks.scheduleRun` calls. Either that or `scheduleRun` will need to be implemented such that any internal retries can't possibly cause updates to be applied out-of-order.

Reply:
I implemented the check for whether to do the immediate update in `requestScheduledAlarm()`, since `commitImpl()` isn't the only place we're using it to set alarms later...  We're also doing it in `armAlarmHandler()`, to reschedule alarms that are too early, and I assume that those requests need the same treatment, to avoid similar problems.

Reply:
Added test case and implementation in ffefa438 .

Reply:
I still think there's something to be said for serializing calls `hooks.scheduleRun` and coalescing queued-up updates together -- we not only have to worry about changes being applied to the scheduler out-of-order, but also about responses from the scheduler coming back out of order and causing us to run the callback that updates `alarmScheduledNoLaterThan` out-of-order.

But this looks good enough to ship 👍